### PR TITLE
topology: dmic: fix pcm_id/dai_link_id confusion

### PR DESCRIPTION
**Pull request by plbossart**
This problem appeared after renumbering the devices for SoundWire
topologies.

Before this patch, the devices appear as:

pcm=JackIn;id=0;type=playback
pcm=JackOut;id=1;type=capture
pcm=DMIC;id=2;type=capture
pcm=DMIC16kHz;id=3;type=capture
pcm=HDMI 1;id=5;type=playback
pcm=HDMI 2;id=6;type=playback
pcm=HDMI 3;id=7;type=playback

and after:

pcm=JackIn;id=0;type=playback
pcm=JackOut;id=1;type=capture
pcm=HDMI 1;id=5;type=playback
pcm=HDMI 2;id=6;type=playback
pcm=HDMI 3;id=7;type=playback
pcm=DMIC;id=10;type=capture
pcm=DMIC16kHz;id=11;type=capture

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]

**Comment by plbossart** [comment]
@singalsu I took your first patch in https://github.com/thesofproject/sof/pull/3841 (only removed the default after agreeing with @ranj063 that this belongs at the higher level).

**Review by ranj063 — APPROVED** [review]

**Comment by plbossart** [comment]
For Pete's sake all topologies are broken now. 

I hate m4!

**Comment by lgirdwood** [comment]
@plbossart looks like a build error
```
m4: /home/sof/work/sof.git/tools/topology/platform/intel/intel-generic-dmic.m4: 16: fatal error: note: Need to define dmic48k pcm id for intel-generic-dmic

make[2]: *** [topology/sof-hda-generic-idisp-4ch.conf] Error 1
make[2]: *** Deleting file 'topology/sof-hda-generic-idisp-4ch.conf'
topology/CMakeFiles/topology_sof-hda-generic-idisp-4ch.dir/build.make:64: recipe for target 'topology/sof-hda-generic-idisp-4ch.conf' failed
CMakeFiles/Makefile2:1863: recipe for target 'topology/CMakeFiles/topology_sof-hda-generic-idisp-4ch.dir/all' failed
make[1]: *** [topology/CMakeFiles/topology_sof-hda-generic-idisp-4ch.dir/all] Error 2
```

**Comment by plbossart** [comment]
> @plbossart looks like a build error
> 
> ```
> m4: /home/sof/work/sof.git/tools/topology/platform/intel/intel-generic-dmic.m4: 16: fatal error: note: Need to define dmic48k pcm id for intel-generic-dmic
> 
> make[2]: *** [topology/sof-hda-generic-idisp-4ch.conf] Error 1
> make[2]: *** Deleting file 'topology/sof-hda-generic-idisp-4ch.conf'
> topology/CMakeFiles/topology_sof-hda-generic-idisp-4ch.dir/build.make:64: recipe for target 'topology/sof-hda-generic-idisp-4ch.conf' failed
> CMakeFiles/Makefile2:1863: recipe for target 'topology/CMakeFiles/topology_sof-hda-generic-idisp-4ch.dir/all' failed
> make[1]: *** [topology/CMakeFiles/topology_sof-hda-generic-idisp-4ch.dir/all] Error 2
> ```

Yes. It's a multi-level bug:
a) the dailink ID was used instead of the PCM ID. That was fixed in a previous commit
b) That probably never worked because the PCM ID was never defined....
c) we added error checks
d) so now we fail on PCM ID.

The solution is to declare both dailink ID and PCM ID in callers.

**Comment by plbossart** [comment]
Compilation *should* work now but let's do a round of validation. I want to double-check the topology binaries for HDaudio platforms are not changed by my changes.

**Comment by singalsu** [comment]
As a check I usually do a diff -udr for master branch topology build .conf vs. new topology build .conf. If only white space changes seen then topologies should remain unmodified.

**Comment by plbossart** [comment]
> As a check I usually do a diff -udr for master branch topology build .conf vs. new topology build .conf. If only white space changes seen then topologies should remain unmodified.

That seems quite wise indeed: I just found that some old Chrome topologies don't rely on the relevant macros and added 'PCM_NAME" in the .conf results.

I compared the topologies with commit  86abd07cbacba10bd44f66e1dfa419694959f430 and the only real changes outside of spaces and comments are for SoundWire topologies (desired) and renaming of HDaudio DMIC48kHz->DMIC.

@ranj063 do you mind taking a look at the last commit, I see some sort of infinite loop when I use 'DMIC16k'. Using DMIC16kHz seems to work. I really have no idea what M4 does. Thanks!

File: tools/topology/sof-cml-rt5682-kwd.m4 [hunk 1]
```diff
@@ -29,6 +29,9 @@ ifdef(`HSEARPROC', , `define(HSEARPROC, volume)')
 ifdef(`DMICPROC', , `define(DMICPROC, passthrough)')
 ifdef(`DMIC16KPROC', , `define(DMIC16KPROC, passthrough)')
 
+# FIXME: Using DMIC16kHz instead of DMIC16k, otherwise M4 does not return.
+define(DMIC_16k_PCM_NAME, DMIC16kHz)
```
Comment (ranj063) [inline tools/topology/sof-cml-rt5682-kwd.m4:33]: @plbossart why do we need this at all? the reason you're running into issues is likely that fact that pipe-kfbm-capture.m4 already adds a PCM called DMIC16k.

Comment (plbossart) [reply tools/topology/sof-cml-rt5682-kwd.m4:33]: if I don't have this then the .conf file has a 'PCM_NAME' difference with the original, the macro is not defined and is taken as text.

**Comment by lgirdwood** [comment]
Device test all green ! @plbossart I assume this is for v1,7 ?

**Comment by plbossart** [comment]
> Device test all green ! @plbossart I assume this is for v1,7 ?

yes we want 1.7, and this also solves our CI issues with the wrong devices being used.

**Pull request merged**